File: src/preprocessing/Filtering.jl
# NeuroMIND/src/preprocessing/Filtering.jl
#
# Cadena de filtrado con soporte para dos perfiles:
#
#   profile = "eeg_julia"  (reproducibilidad estricta con EEG_Julia)
#     Orden:  Notch(filt) → Bandreject(filt) → Highpass(filtfilt) → Lowpass(filtfilt)
#     Refs:   ../EEG_Julia/src/Preprocessing/filtering.jl
#             ../EEG_Julia/config/default_config.jl
#
#   profile = "default"  (legacy NeuroMIND / protocolo BrainVision)
#     Orden:  Highpass(filtfilt) → Lowpass(filtfilt) → Notch(filtfilt) → Bandreject(filtfilt)
#
# El perfil activo se lee de cfg.filtering["profile"].
# Si la clave no existe se asume "default".

"""
    filter_recording(rec, cfg) -> EEGRecording

Aplica la cadena completa de filtrado según `cfg.filtering["profile"]`.

Perfil `"eeg_julia"` (reproducibilidad estricta):
  1. Notch      50 Hz, bw 1 Hz, order 4, método: `filt` (causal)
  2. Bandreject 100 Hz, bw 1 Hz → (99.5–100.5 Hz), order 4, método: `filt` (causal)
  3. High-pass   0.5 Hz, order 4, método: `filtfilt` (zero-phase)
  4. Low-pass  150.0 Hz, order 4, método: `filtfilt` (zero-phase)

Perfil `"default"` (legacy):
  1. High-pass  (filtfilt)
  2. Low-pass   (filtfilt)
  3. Notch      (filtfilt, si notch_hz > 0)
  4. Bandreject (filtfilt, si bandreject_lo > 0 && hi > lo)
"""
function filter_recording(rec::EEGRecording, cfg::PipelineConfig)::EEGRecording
    f       = cfg.filtering
    data    = copy(rec.data)
    fs      = rec.meta.fs
    profile = get(f, "profile", "default")
    ord     = Int(get(f, "filter_order", 4))
    hp      = Float64(get(f, "highpass_hz",    0.5))
    lp      = Float64(get(f, "lowpass_hz",    48.0))
    nz      = Float64(get(f, "notch_hz",      50.0))
    nbw     = Float64(get(f, "notch_bw_hz",    2.0))
    lo      = Float64(get(f, "bandreject_lo",  0.0))
    hi      = Float64(get(f, "bandreject_hi",  0.0))

    if profile == "eeg_julia"
        # ── Reproducibilidad EEG_Julia ─────────────────────────
        # Paso 1 — Notch causal (filt, order 4, bw 1 Hz)
        nz > 0.0 && (data = _filt(data, fs, :notch;    freq=nz, bw=nbw,       order=ord, method=:filt))
        # Paso 2 — Bandreject causal (filt, order 4, 99.5–100.5 Hz)
        lo > 0.0 && hi > lo && (data = _filt(data, fs, :bandstop; freq=(lo,hi), order=ord, method=:filt))
        # Paso 3 — Highpass zero-phase (filtfilt, order 4)
        data = _filt(data, fs, :highpass; freq=hp, order=ord, method=:filtfilt)
        # Paso 4 — Lowpass zero-phase (filtfilt, order 4)
        data = _filt(data, fs, :lowpass;  freq=lp, order=ord, method=:filtfilt)
    else
        # ── Protocolo BrainVision / legacy ─────────────────────
        data = _filt(data, fs, :highpass; freq=hp,  order=ord)
        data = _filt(data, fs, :lowpass;  freq=lp,  order=ord)
        nz > 0.0 && (data = _filt(data, fs, :notch; freq=nz, bw=nbw))
        lo > 0.0 && hi > lo &&
            (data = _filt(data, fs, :bandstop; freq=(lo, hi), order=ord))
    end

    return EEGRecording(rec.meta, data, rec.times)
end

"""
    describe_filter_chain(cfg) -> Vector{NamedTuple}

Devuelve la cadena de filtros que aplicará `filter_recording` según `cfg.filtering["profile"]`.
Cada elemento: `(step, name, freq, order, method)`.

Útil para logging, dashboard y validación.
"""
function describe_filter_chain(cfg::PipelineConfig)
    f       = cfg.filtering
    profile = get(f, "profile", "default")
    ord     = Int(get(f, "filter_order", 4))
    hp      = Float64(get(f, "highpass_hz",    0.5))
    lp      = Float64(get(f, "lowpass_hz",    48.0))
    nz      = Float64(get(f, "notch_hz",      50.0))
    nbw     = Float64(get(f, "notch_bw_hz",    2.0))
    lo      = Float64(get(f, "bandreject_lo",  0.0))
    hi      = Float64(get(f, "bandreject_hi",  0.0))

    if profile == "eeg_julia"
        chain = NamedTuple[]
        step  = 1
        nz > 0.0 && (push!(chain, (step=step, name="Notch",
            freq="$(nz - nbw/2)–$(nz + nbw/2) Hz",
            order=ord, method="filt")); step += 1)
        lo > 0.0 && hi > lo && (push!(chain, (step=step, name="Bandreject",
            freq="$(lo)–$(hi) Hz",
            order=ord, method="filt")); step += 1)
        push!(chain, (step=step, name="High-pass",
            freq="$(hp) Hz",   order=ord, method="filtfilt")); step += 1
        push!(chain, (step=step, name="Low-pass",
            freq="$(lp) Hz",   order=ord, method="filtfilt"))
        return chain
    else
        chain = NamedTuple[
            (step=1, name="High-pass", freq="$(hp) Hz", order=ord, method="filtfilt"),
            (step=2, name="Low-pass",  freq="$(lp) Hz", order=ord, method="filtfilt"),
        ]
        s = 3
        nz > 0.0 && (push!(chain, (step=s, name="Notch",
            freq="$(nz - nbw/2)–$(nz + nbw/2) Hz",
            order=ord, method="filtfilt")); s += 1)
        lo > 0.0 && hi > lo && push!(chain, (step=s, name="Bandreject",
            freq="$(lo)–$(hi) Hz",
            order=ord, method="filtfilt"))
        return chain
    end
end

# ─── Filtros individuales (API pública) ───────────────────────

function apply_highpass(rec::EEGRecording, cutoff_hz::Real; order::Int=4)::EEGRecording
    EEGRecording(rec.meta, _filt(rec.data, rec.meta.fs, :highpass; freq=cutoff_hz, order=order), rec.times)
end

function apply_lowpass(rec::EEGRecording, cutoff_hz::Real; order::Int=4)::EEGRecording
    EEGRecording(rec.meta, _filt(rec.data, rec.meta.fs, :lowpass; freq=cutoff_hz, order=order), rec.times)
end

function apply_notch(rec::EEGRecording, center_hz::Real, bw_hz::Real=2.0)::EEGRecording
    EEGRecording(rec.meta, _filt(rec.data, rec.meta.fs, :notch; freq=center_hz, bw=bw_hz), rec.times)
end

function apply_bandpass(rec::EEGRecording, lo::Real, hi::Real; order::Int=8)::EEGRecording
    EEGRecording(rec.meta, _filt(rec.data, rec.meta.fs, :bandpass; freq=(lo, hi), order=order), rec.times)
end

function apply_bandreject(rec::EEGRecording, lo::Real, hi::Real; order::Int=4)::EEGRecording
    EEGRecording(rec.meta, _filt(rec.data, rec.meta.fs, :bandstop; freq=(lo, hi), order=order), rec.times)
end

# ─── Kernel interno ───────────────────────────────────────────

"""
    _filt(data, fs, type; freq, bw, order, method) -> Matrix{Float64}

Kernel de filtrado sobre todas las filas (canales) de una matriz.
`method = :filtfilt` (zero-phase, por defecto) | `:filt` (causal, compatible EEG_Julia).
"""
function _filt(data::Matrix{Float64}, fs::Real, type::Symbol;
               freq=nothing, bw::Real=2.0, order::Int=4,
               method::Symbol=:filtfilt)::Matrix{Float64}
    flt = _build(type, fs, freq, bw, order)
    out = similar(data)
    @inbounds for ch in 1:size(data, 1)
        out[ch, :] = method === :filt ? filt(flt, data[ch, :]) : filtfilt(flt, data[ch, :])
    end
    return out
end

function _build(type::Symbol, fs::Real, freq, bw::Real, order::Int)
    nyq = fs / 2.0
    if type === :highpass
        return digitalfilter(Highpass(Float64(freq) / nyq), Butterworth(order))
    elseif type === :lowpass
        return digitalfilter(Lowpass(Float64(freq) / nyq), Butterworth(order))
    elseif type === :bandpass
        return digitalfilter(Bandpass(Float64(freq[1]) / nyq, Float64(freq[2]) / nyq), Butterworth(order))
    elseif type === :notch || type === :bandstop
        lo = isa(freq, Tuple) ? Float64(freq[1]) / nyq : (Float64(freq) - bw / 2) / nyq
        hi = isa(freq, Tuple) ? Float64(freq[2]) / nyq : (Float64(freq) + bw / 2) / nyq
        return digitalfilter(Bandstop(lo, hi), Butterworth(order))
    else
        error("Tipo de filtro desconocido: $type")
    end
end

File: mne_brain/src/mne_brain/preprocessing/filtering.py
"""Butterworth filtering compatible with NeuroMIND's `eeg_julia` profile."""

from __future__ import annotations

from typing import Any

import numpy as np

from mne_brain.common.types import EEGRecording, PipelineConfig


def filter_recording(rec: EEGRecording, cfg: PipelineConfig) -> EEGRecording:
    """Apply the configured NeuroMIND filter chain to an EEG recording.

    For `profile = "eeg_julia"` the exact intended order is:

    1. 50 Hz notch, causal `sosfilt`
    2. 99.5-100.5 Hz bandreject, causal `sosfilt`
    3. 0.5 Hz high-pass, zero-phase `sosfiltfilt`
    4. 150 Hz low-pass, zero-phase `sosfiltfilt`
    """
    filt_cfg = cfg.filtering
    data = np.array(rec.data, dtype=np.float64, copy=True)
    fs = float(rec.meta.fs)
    profile = str(filt_cfg.get("profile", "default"))
    order = int(filt_cfg.get("filter_order", 4))
    hp = float(filt_cfg.get("highpass_hz", 0.5))
    lp = float(filt_cfg.get("lowpass_hz", 150.0))
    notch = float(filt_cfg.get("notch_hz", 50.0))
    notch_bw = float(filt_cfg.get("notch_bw_hz", 1.0))
    br_lo = float(filt_cfg.get("bandreject_lo", 0.0))
    br_hi = float(filt_cfg.get("bandreject_hi", 0.0))

    if profile == "eeg_julia":
        if notch > 0.0:
            data = _apply_sos(
                data,
                fs,
                "bandstop",
                (notch - notch_bw / 2, notch + notch_bw / 2),
                order,
                method="filt",
            )
        if br_lo > 0.0 and br_hi > br_lo:
            data = _apply_sos(data, fs, "bandstop", (br_lo, br_hi), order, method="filt")
        data = _apply_sos(data, fs, "highpass", hp, order, method="filtfilt")
        data = _apply_sos(data, fs, "lowpass", lp, order, method="filtfilt")
    else:
        data = _apply_sos(data, fs, "highpass", hp, order, method="filtfilt")
        data = _apply_sos(data, fs, "lowpass", lp, order, method="filtfilt")
        if notch > 0.0:
            data = _apply_sos(
                data,
                fs,
                "bandstop",
                (notch - notch_bw / 2, notch + notch_bw / 2),
                order,
                method="filtfilt",
            )
        if br_lo > 0.0 and br_hi > br_lo:
            data = _apply_sos(data, fs, "bandstop", (br_lo, br_hi), order, method="filtfilt")

    return EEGRecording(meta=rec.meta, data=data, times=rec.times.copy())


def describe_filter_chain(cfg: PipelineConfig) -> list[dict[str, Any]]:
    """Return the filter chain as dashboard-friendly dictionaries."""
    filt_cfg = cfg.filtering
    profile = str(filt_cfg.get("profile", "default"))
    order = int(filt_cfg.get("filter_order", 4))
    hp = float(filt_cfg.get("highpass_hz", 0.5))
    lp = float(filt_cfg.get("lowpass_hz", 150.0))
    notch = float(filt_cfg.get("notch_hz", 50.0))
    notch_bw = float(filt_cfg.get("notch_bw_hz", 1.0))
    br_lo = float(filt_cfg.get("bandreject_lo", 0.0))
    br_hi = float(filt_cfg.get("bandreject_hi", 0.0))

    chain: list[dict[str, Any]] = []
    if profile == "eeg_julia":
        if notch > 0.0:
            chain.append(
                _step(
                    len(chain) + 1,
                    "Notch",
                    f"{notch - notch_bw / 2}-{notch + notch_bw / 2} Hz",
                    order,
                    "filt",
                )
            )
        if br_lo > 0.0 and br_hi > br_lo:
            chain.append(_step(len(chain) + 1, "Bandreject", f"{br_lo}-{br_hi} Hz", order, "filt"))
        chain.append(_step(len(chain) + 1, "High-pass", f"{hp} Hz", order, "filtfilt"))
        chain.append(_step(len(chain) + 1, "Low-pass", f"{lp} Hz", order, "filtfilt"))
        return chain

    chain.append(_step(1, "High-pass", f"{hp} Hz", order, "filtfilt"))
    chain.append(_step(2, "Low-pass", f"{lp} Hz", order, "filtfilt"))
    if notch > 0.0:
        chain.append(
            _step(
                len(chain) + 1,
                "Notch",
                f"{notch - notch_bw / 2}-{notch + notch_bw / 2} Hz",
                order,
                "filtfilt",
            )
        )
    if br_lo > 0.0 and br_hi > br_lo:
        chain.append(_step(len(chain) + 1, "Bandreject", f"{br_lo}-{br_hi} Hz", order, "filtfilt"))
    return chain


def apply_highpass(rec: EEGRecording, cutoff_hz: float, *, order: int = 4) -> EEGRecording:
    data = _apply_sos(rec.data, rec.meta.fs, "highpass", cutoff_hz, order, method="filtfilt")
    return EEGRecording(rec.meta, data, rec.times.copy())


def apply_lowpass(rec: EEGRecording, cutoff_hz: float, *, order: int = 4) -> EEGRecording:
    data = _apply_sos(rec.data, rec.meta.fs, "lowpass", cutoff_hz, order, method="filtfilt")
    return EEGRecording(rec.meta, data, rec.times.copy())


def apply_notch(
    rec: EEGRecording,
    center_hz: float,
    bw_hz: float = 2.0,
    *,
    order: int = 4,
) -> EEGRecording:
    band = (center_hz - bw_hz / 2, center_hz + bw_hz / 2)
    data = _apply_sos(rec.data, rec.meta.fs, "bandstop", band, order, method="filtfilt")
    return EEGRecording(rec.meta, data, rec.times.copy())


def apply_bandreject(rec: EEGRecording, lo: float, hi: float, *, order: int = 4) -> EEGRecording:
    data = _apply_sos(rec.data, rec.meta.fs, "bandstop", (lo, hi), order, method="filtfilt")
    return EEGRecording(rec.meta, data, rec.times.copy())


def _apply_sos(
    data: np.ndarray,
    fs: float,
    kind: str,
    freq: float | tuple[float, float],
    order: int,
    *,
    method: str,
) -> np.ndarray:
    from scipy.signal import butter, sosfilt, sosfiltfilt

    btype = {
        "highpass": "highpass",
        "lowpass": "lowpass",
        "bandpass": "bandpass",
        "bandstop": "bandstop",
    }[kind]
    sos = butter(order, freq, btype=btype, fs=fs, output="sos")
    if method == "filt":
        return sosfilt(sos, data, axis=1)
    if method == "filtfilt":
        return sosfiltfilt(sos, data, axis=1)
    raise ValueError(f"Unknown filter method: {method}")


def _step(step: int, name: str, freq: str, order: int, method: str) -> dict[str, Any]:
    return {
        "step": step,
        "name": name,
        "type": "Butterworth",
        "freq": freq,
        "order": order,
        "method": method,
        "applied": True,
    }
